### PR TITLE
Change the boundary value type to match the vector type.

### LINKED ISSUE #5417: Add complex-valued tests
Now that we instantiate everything for complex-valued vectors (see #2033), we ought to add a few tests for at least the common cases. Candidates would be:
- [x] `VectorTools::interpolate`
- [ ] `KellyErrorEstimator::estimate`

`DataOut` will be tested as part of #1894.

### PR DESCRIPTION
**Pull request by bangerth**
Fixes #5417.

### CODE REVIEW

**Comment by bangerth** [comment]
I should have been more explicit: this changes the type of the Neumann boundary values given to `KellyErrorEstimator::estimate()`. I needed to do this because that is clearly the right thing. Noticed while working on a testcase for #5417.

File: doc/news/changes/incompatibilities/20171117Bangerth [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Changed: The many KellyErrorEstimator::estimate() overloads all took
```
Comment (jppelteret) [inline doc/news/changes/incompatibilities/20171117Bangerth]: all took -> all taken

File: doc/news/changes/incompatibilities/20171117Bangerth [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+function arguments (or FunctionMap arguments) indicating the Neumann
```
Comment (jppelteret) [inline doc/news/changes/incompatibilities/20171117Bangerth]: Can you please explain why have you described this as 

> function arguments (or FunctionMap arguments)

? As far as I can ell, all of the changed arguments are of type `FunctionMap`.  Maybe this relates to my next question below...

File: include/deal.II/numerics/error_estimator.templates.h [hunk 3]
```diff
@@ -169,9 +169,9 @@ namespace internal
        * Some more references to input data to the
        * KellyErrorEstimator::estimate() function.
        */
-      const typename FunctionMap<spacedim>::type *neumann_bc;
-      const ComponentMask                component_mask;
-      const Function<spacedim>                   *coefficients;
+      const typename FunctionMap<spacedim,number>::type *neumann_bc;
+      const ComponentMask                                component_mask;
+      const Function<spacedim>                          *coefficients;
```
Comment (jppelteret) [inline include/deal.II/numerics/error_estimator.templates.h]: I admit to being completely naïve wrt. how this function works, but can one not now have complex `coefficients`? ie. Should it be `Function<spacedim>` -> `Function<spacedim,number>` here and below?

File: doc/news/changes/incompatibilities/20171117Bangerth (hunk 1 above)
Comment (luca-heltai) [reply doc/news/changes/incompatibilities/20171117Bangerth]: mmmm... I must say I agree with Wolfgang selection of `took` vs `taken`, i.e., `the overloads took as argument`.  `all taken` does not seem to be right w.r.t. what follows...

Comment (jppelteret) [reply doc/news/changes/incompatibilities/20171117Bangerth]: Whoops. I think I meant to suggest `all take`

Comment (jppelteret) [reply doc/news/changes/incompatibilities/20171117Bangerth]: But now reading further, I submit that even that is wrong because the function has had its arguments changed. Sorry, and please ignore!

Comment (jppelteret) [reply doc/news/changes/incompatibilities/20171117Bangerth]: Actually, as far as I can tell all of the changes in the function signatures are, to date,
`typename FunctionMap<spacedim>::type` -> `typename FunctionMap<spacedim,typename InputVector::value_type>::type`. So `all take` is correct.

File: doc/news/changes/incompatibilities/20171117Bangerth [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+of the input vector. In other words, if you are using a regular
+Vector<double> (or PETSc or Trilinos equivalents), then nothing has
+changed since @p number is @p double, which is the default for the
+Function class. On the other hand, if you are using a
+Vector<std::complex<double>>, for example, then you need to pass a
+Function<dim,std::complex<double>> -- which of course makes sense
+because the Neumann boundary values of a complex-valued solution
+should also be complex-valued.
+<br>
+(Wolfgang Bangerth, 2017/09/12)
```
Comment (jppelteret) [inline doc/news/changes/incompatibilities/20171117Bangerth]: Oh, I've also just noticed that this date is incorrect :-)

**Comment by bangerth** [comment]
I've reworded the changelog entry, hopefully making it easier to understand.

@jppelteret -- about your question on why the coefficient in the equation has not been converted: I thought that I had written that down somewhere else already, but can't find it, so here goes:

The Kelly error estimator was originally intended for the Laplace equation, or more generally elliptic equations. For these, to be elliptic, the coefficient needs to be Hermitian. If the coefficient was a matrix, then that matrix would need to be Hermitian. But the class assumes that the coefficient is in fact a scalar function (or maybe a scalar for each solution variable, so a diagonal matrix). A Hermitian scalar, or a Hermitian diagonal matrix, is necessarily real. So I didn't see much of a use case for allowing a complex-valued coefficient.

I do recognize that the class is used in contexts outside its original intent. But I thought we can go down this road when someone demonstrates a need. I don't see that need right now.

**Comment by jppelteret** [comment]
Thanks as well for the explanation wrt. the `coefficients`. Very interesting, and I learned something new about error estimators!

**Comment by bangerth** [comment]
Part of #2033.
